### PR TITLE
proxy+pageserver: shared leaky bucket impl

### PR DESCRIPTION
**Pull request by conradludgate**
In proxy I switched to a leaky-bucket impl using the GCRA algorithm. I figured I could share the code with pageserver and remove the leaky_bucket crate dependency with some very basic tokio timers and queues for fairness.

The underlying algorithm should be fairly clear how it works from the comments I have left in the code.

---

In benchmarking pageserver, @problame found that the new implementation fixes a getpage throughput discontinuity in pageserver under the `pagebench get-page-latest-lsn` benchmark with the clickbench dataset (`test_perf_olap.py`).
The discontinuity is that for any of `--num-clients={2,3,4}`, getpage throughput remains 10k.
With `--num-clients=5` and greater, getpage throughput then jumps to the configured 20k rate limit.
With the changes in this PR, the discontinuity is gone, and we scale throughput linearly to `--num-clients` until the configured rate limit.

More context in https://github.com/neondatabase/cloud/issues/16886#issuecomment-2315257641.

closes https://github.com/neondatabase/cloud/issues/16886

### CODE REVIEW

**Comment by conradludgate** [comment]
Upon testing, it seems the current leaky-bucket crate supports waiting for more tokens than can ever fit in the bucket. This isn't working properly with my impl just yet (it technically still works but it first resets the bucket to be empty, which breaks the time tracking, so it ends up waiting longer than it should)

**Comment by koivunej** [comment]
> I'm not sure if this feature is important for ps

I think all of the waits are for one token, so no.

File: libs/utils/src/leaky_bucket.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+    /// We track all times as durations since this epoch so we can round down.
+    pub epoch: Instant,
+
+    /// How frequently we drain the bucket.
+    /// If equal to 0, we drain continuously over time.
+    /// If greater than 0, we drain at fixed intervals.
+    pub drain_interval: Duration,
+
+    /// "time cost" of a single request unit.
+    /// should loosely represents how long it takes to handle a request unit in active resource time.
```
Comment (koivunej) [inline libs/utils/src/leaky_bucket.rs:16]:
```suggestion
    /// should loosely represent how long it takes to handle a request unit in active resource time.
```

**Comment by conradludgate** [comment]
It should now be 100% compatible by tracking the start time, and only adjusting empty bucket position based on that start time.

**Comment by problame** [comment]
I just resolved the conflicts by merging from `main`.

See commit message for how I resolved conflicts: https://github.com/neondatabase/neon/pull/8539/commits/5b9d37151bc0f6d239dcf5489ce91d096cfee024

**Comment by problame** [comment]
Context why I'm looking into this: https://github.com/neondatabase/cloud/issues/16886#issuecomment-2315257641

=> @conradludgate please review my recent pushes and let's get this merged.

Here's my proposed updated PR description

```
[EXISITING TEXT]

In benchmarking pageserver, @problame found that the new implementation fixes a getpage throughput discontinuity in pageserver under the `pagebench get-page-latest-lsn` benchmark with the clickbench dataset (`test_perf_olap.py`).
The discontinuity is that for any of `--num-clients={2,3,4}`, getpage throughput remains 10k.
With `--num-clients=5` and greater, getpage throughput then jumps to the configured 20k rate limit.
With the changes in this PR, the discontinuity is gone, and we scale throughput linearly to `--num-clients` until the configured rate limit.

More context in https://github.com/neondatabase/cloud/issues/16886#issuecomment-2315257641.

closes https://github.com/neondatabase/cloud/issues/16886
```

File: pageserver/src/tenant/throttle.rs [hunk 2]
```diff
@@ -93,18 +94,34 @@ where
                 }
             })
             .collect();
+
+        // how frequently we drain a single token on average
+        let time_cost = refill_interval / refill_amount.get() as u32;
```
Comment (problame) [inline pageserver/src/tenant/throttle.rs:99]: This conversion to u32 is lossy, let's not do that. Will think of something.

File: libs/utils/src/leaky_bucket.rs [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+        if self.drain_interval > Duration::ZERO {
+            let n = dur.div_duration_f64(self.drain_interval).ceil();
+            dur = self.drain_interval.mul_f64(n);
+        }
+        dur
+    }
+}
+
+pub struct LeakyBucketState {
+    /// Bucket is represented by `start..end` where `end = epoch + end` and `start = end - config.bucket_width`.
```
Comment (problame) [inline libs/utils/src/leaky_bucket.rs:42]: `end = epoch + end`  ?

File: libs/utils/src/leaky_bucket.rs [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+    /// Draining the bucket will happen naturally as `now` moves forward.
+    ///
+    /// Let `n` be some "time cost" for the request,
+    /// If now is after end, the bucket is empty and the end is reset to now,
+    /// If now is within the `bucket window + n`, we are within time budget.
+    /// If now is before the `bucket window + n`, we have run out of budget.
+    ///
+    /// This is inspired by the generic cell rate algorithm (GCRA) and works
+    /// exactly the same as a leaky-bucket.
+    pub end: Duration,
```
Comment (problame) [inline libs/utils/src/leaky_bucket.rs:56]: I feel like this doc comment should be at the top of the module. Without it, the module is a lot harder to understand. Put it in front of the reader's eyes before they dive into the code.

Also, it's not explained anywhere where `start` comes from.

Mention of GCRA and a link to the wikipedia page should IMO be in the first pargagraph, not the last.

File: libs/utils/src/leaky_bucket.rs [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+        // If the bucket was empty when we started our search, bump the end up accordingly.
+        let mut end = self.end;
+        if end < started {
+            end = started;
+        }
+
+        let n = config.cost.mul_f64(n);
+        let end_plus_n = end + n;
+        let start_plus_n = end_plus_n.saturating_sub(config.bucket_width);
```
Comment (problame) [inline libs/utils/src/leaky_bucket.rs:103]: This variable name is super confusing to me. It suggests this is `start + n` but it actually is more like "when would we have needed to start to stay exactly within the budget limit". No?

Comment (problame) [reply libs/utils/src/leaky_bucket.rs:103]:
```suggestion
        let wait_until = end_plus_n.saturating_sub(config.bucket_width);
```

**Review by problame — COMMENTED** [review]
reviewed `libs/utils/src/leaky_bucket.rs`. Didn't know GCRA before and only skimmed the Wikipedia article. Some comments

**Comment by problame** [comment]
Call with conrad:

1. remove the discrete refill - pageserver doesn't rely on it
2. make `queue` mandatory
3. move RateLimiter into shared crate

File: pageserver/src/tenant/throttle.rs (hunk 2 above)
Comment (conradludgate) [reply pageserver/src/tenant/throttle.rs]: updated the config to use u32s

File: libs/pageserver_api/src/models.rs [hunk 6]
```diff
@@ -486,11 +486,11 @@ pub struct EvictionPolicyLayerAccessThreshold {
...
     pub task_kinds: Vec<String>, // TaskKind
-    pub initial: usize,
+    pub initial: u32,
     #[serde(with = "humantime_serde")]
     pub refill_interval: Duration,
-    pub refill_amount: NonZeroUsize,
-    pub max: usize,
+    pub refill_amount: NonZeroU32,
+    pub max: u32,
     pub fair: bool,
```
Comment (problame) [inline libs/pageserver_api/src/models.rs:494]: This field is now ignored - the new implementation is always fair.

The field _is_ set in the production pageserver configs.

But we don't have deny_unknown_fields set (a blunder!) so I think we can remov the `fair` field in this PR, wait for it to roll out,t hen remove the `fair` field from the production configs.

**Review by problame — APPROVED** [review]
Did a perf test after the recent changes pushed by Conrad. All looking good.

**Review by cloneable — APPROVED** [review]
LGTM

Not sure why state and config are kept completely separate instead of keeping both inside a LeakyBucket type.

**Comment by conradludgate** [comment]
> Not sure why state and config are kept completely separate instead of keeping both inside a LeakyBucket type.

In proxy, we have 1 config globally (32 bytes), then 1 state per endpoint (16 bytes), hence the split

**Comment by conradludgate** [comment]
I fixed the issue regardless, so it should not matter

**Comment by koivunej** [comment]
I was replying on the basis of page_service requests done which I think are the only throttled ones. I think my answer still holds. Please let me know if that is wrong.

**Comment by Bodobolero** [comment]
Dashboard still doesn't show the 44 second elapsed time for query-1 that we had before (see green line in )

[benchmark history](https://neonprod.grafana.net/d/7XfoNGKVk/perf-test3a-olap?orgId=1&var-test_name=All&var-agg_func=avg&var-benchmark=test_clickbench&viewPanel=46&from=1722439977985&to=1725031977985)

**Comment by problame** [comment]
@Bodobolero let's continue the discussion in the investigation issue https://github.com/neondatabase/cloud/issues/16886#issuecomment-2324742299

**Pull request merged**